### ISSUE TITLE
[feat] Allow use of downlevelIteration on es2015 or greater

### ISSUE THREAD

**Issue by Toxicable** [Microsoft/TypeScript#25025]
## Search Terms
downlevelIteration es2015 es6

## Suggestion
A new flag or way to enable `downlevelIteration` when targeting es2015 or greater
Maybe `forceDownlevelIteration` ?

## Use Cases
`for .. of` is slow: https://jsperf.com/for-vs-forof
This request comes from my recent work in Nodejs where you want to use higher targets for better performance from the runtime but have to replace `for .. of` with `for` due to it's inefficiencies.
Devs shouldn't be forced to write less ergonomic code for performance reasons when TypeScript could simply do it for you.

An example of having to refactor due to increasing the language target:
https://github.com/angular/angular/pull/24534

## Examples
N/A

## Checklist

My suggestion meets these guidelines:
* [x] This wouldn't be a breaking change in existing TypeScript / JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. new expression-level syntax)

**Comment by mhegazy** [comment]
Historically, we have not allowed such granular transformations.

**Comment by Toxicable** [comment]
@DanielRosenwasser I don't follow.
I would expect that an option was added along the lines of `forDownlevelIteration` that would produce the same code for the example below in es5 and es2015
```
var map = new Map<number, string>();
for(let item of map.keys()) { }
```

**Comment by mhegazy** [comment]
The underlying issue here seems to be one about engine performance. I do not see why this has to be a permanent issue, V8/Chakra/JSC/SpiderMonkey could be better are optimizing these patterns in the future, and that solves the issue.
Now for the request in the OP, as i noted earlier, we have not supported such level of configrability for our transforms. we have found that for most users transforming to a specific target language covers their needs; and adding extra configuration and maintenance complexity does not introduce much added value.

**Comment by Toxicable** [comment]
@mhegazy I agree with the first part. It would be great if this wasn't an issue in the first place.
As for the second part; I disagree how much value this would provide.
Right now people are moving to higher targets for better performance but at the same time hurting their performance, possibly without even knowing how bad `for .. of` is.
As im sure you know the vast majority of execution time in programs that need high performance is spent in loops.

**Comment by kitsonk** [comment]
There is nothing preventing people writing the target themselves for performance reasons.  Much as many people will not use a lot of the Array extras in code (e.g. `.forEach()`) because of performance concerns.

Even optionally, it feels to me like it goes against the design principles of TypeScript, where code is only rewritten to meet the target.  People could use static analysis to enforce a coding choice, either through a linter, or using other general JavaScript to JavaScript compilers to improve performance.

**Comment by Toxicable** [comment]
@kitsonk Agreed, but as I pointed out above, not everyone knows that most of the other methods of iteration are slow compared to `for`.

Also writing `for` loops are much less ergonomic compared to `for ... of`, I personally would like to use the newer language features without feeling hindered by the performance concerns that go with it.

**Comment by kitsonk** [comment]
> Agreed, but as I pointed out above, not everyone knows that most of the other methods of iteration are slow compared to `for`.

So people have to _know_ to enable the flag, but can be ignorant of performance.  There are certainly value for those type of tools, but IMO TypeScript is not that tool.

> I personally would like to use the newer language features without feeling hindered by the performance concerns that go with it

Your assumption is that all runtimes are created equal and they are static and don't improve the performance of language features.  You are then projecting that into your transpiler.  Again, that might be suitable for you, but it is a very slippery slope that I don't think TypeScript should get into.  But again that is just my opinion.

**Comment by alfaproject** [comment]
Sounds to me that what you want is a TSLint rule or something along those lines. It's not TS responsibility to downlevel language features that are natively supported by the target. That goes against its philosophy, surely.

**Comment by RyanCavanaugh** [comment]
Declining since we don't really want to keep furthering any runtime differences between TS and JS
